File: Stack/232.Implement-Queue-Using-Stacks/232.Implement-Queue-Using-Stacks.ts
class MyQueue {
  private insertStack: number[];
  private deleteStack: number[];
  constructor() {
    this.insertStack = [];
    this.deleteStack = [];
  }

  push(x: number): void {
    this.insertStack.push(x);
  }

  pop(): number {
    if (this.deleteStack.length > 0) {
      return this.deleteStack.pop();
    } else {
      this.move();
      return this.deleteStack.pop();
    }
  }

  private move() {
    while (this.insertStack.length) {
      this.deleteStack.push(this.insertStack.pop());
    }
  }

  peek(): number {
    if (this.deleteStack.length === 0) {
      this.move();
    }
    return this.deleteStack[this.deleteStack.length - 1];
  }

  empty(): boolean {
    return this.insertStack.length === 0 && this.deleteStack.length === 0;
  }
}

File: Stack/232.Implement-Queue-Using-Stacks/232.Implement-Queue-Using-Stacks.py
class MyQueue:

    def __init__(self):
        self.insertStack = []
        self.deleteStack = []


    def push(self, x: int) -> None:
        self.insertStack.append(x)

    def pop(self) -> int:
        self.moveIfdeleteStackEmpty()
        return self.deleteStack.pop()


    def peek(self) -> int:
        self.moveIfdeleteStackEmpty()
        return self.deleteStack[-1]


    def empty(self) -> bool:
        return len(self.insertStack) == 0 and len(self.deleteStack) == 0
    
    def moveIfdeleteStackEmpty(self) -> None:
        if len(self.deleteStack) == 0:
            while len(self.insertStack) > 0:
                self.deleteStack.append(self.insertStack.pop())
                



# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()
